### DISCUSSION TITLE
Where to find loading direction in source codes

### DISCUSSION THREAD

**Discussion by pshen20127** [idaholab/moose#23341] [Q&A Modules: Solid mechanics] [answered]
Hello All

I need to add some lines in source codes to calculate schmid factor and need loading direction. Do you know which variable in the source codes is loading direction? The files I need to modify are 'CrystalPlasticityKalidindiUpdate.C' and 'CrystalPlasticityStressUpdateBase.C'.

Thanks
Pengfei

**Comment by GiudGiud** [comment]
@sapitts

**Comment by sapitts** [comment, marked as answer]
There is not a way to extract the loading direction from the current material properties in `CrystalPlasticityStressUpdateBase.C`. When we calculate the applied shear stress for each slip system, we use full PK2 stress tensor which includes the loading information in the component magnitudes. The rotation tensor calculated from the Euler angles, `_crysrot[_qp]` does not include enough unique information to determine the loading direction, so it is not possible to calculate the scalar Schmid factor from the double contraction of the `_crysrot` and `_flow_direction` tensors. @dewenyushu for a double check here please.

In order to calculate the scalar Schmid factor, I would recommend the following steps:
1. provide the loading direction via an input parameter, making sure that the loading direction is provided in the same crystal sample system in which the slip system directions and plane normals are provided
2. normalize the loading direction vector, see https://github.com/idaholab/moose/blob/11b38779f26d9f5f1abf28849f49ac5214fe9d5a/modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityStressUpdateBase.C#L180
3. Implement the expression $M = (\hat{n} * \hat{d}) * (\hat{m} * \hat{d})$ where $\hat{d}$ is the normalized loading direction vector to calculate the initial Schmid factor
4. If the updated Schmid factor values throughout the system are desired, multiply each of the $\hat{m}$, $\hat{d}$, and $\hat{m}$ vectors by the `_crysrot` tensor, in the manner shown here https://github.com/idaholab/moose/blob/11b38779f26d9f5f1abf28849f49ac5214fe9d5a/modules/tensor_mechanics/src/materials/crystal_plasticity/CrystalPlasticityStressUpdateBase.C#L382-L396
5. Calculate the updated Schmid factor using the updated $\hat{m}$, $\hat{d}$, and $\hat{m}$ vectors in the expression from step 3 above
I would recommend saving each of the slip system Schmid factors in a material property vector of size `_number_slip_systems` so that you can then use the same postprocessors that we use for other slip system material properties like the applied stress

Finally, if you are not planning to make a pull request to MOSOE with these changes, I woudl recommend implementing these changes in a child class that inherits from `CrystalPlasticityKalidindiUpdate.C`. Using this implementation approach will enable you to stay up-to-date with MOOSE by rebasing with `upstream/devel` often.

Hope this helps,
Stephanie

**Reply by dewenyushu** [reply]
I agree with @sapitts 

The loading direction is not defined in the materials classes. On the contrary, the materials models response to the different load directions (and conditions) and initial orientations you may specify in your input file

Adding the loading direction as an input parameter sounds reasonable if you will need it in calculating the Schmid factor

**Reply by hugary1995** [reply]
If you are looking at an RVE, just homogenize the cauchy stress -- that's the "load" you use to compute the CRSS with Schmid.

**Reply by pshen20127** [reply]
Thank you for so detailed reply. How to provide the loading direction via an input parameter?

**Reply by hugary1995** [reply]
https://mooseframework.inl.gov/getting_started/examples_and_tutorials/index.html
